ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/Database.scala

```diff
@@ -3,33 +3,86 @@
 
 package com.daml.ledger.on.sql
 
+import java.io.Closeable
+
 import com.daml.ledger.on.sql.queries.Queries.InvalidDatabaseException
 import com.daml.ledger.on.sql.queries.{H2Queries, Queries, SqliteQueries}
+import com.digitalasset.logging.{ContextualizedLogger, LoggingContext}
+import com.zaxxer.hikari.HikariDataSource
+import javax.sql.DataSource
 
-sealed trait Database {
+sealed trait Database extends Closeable {
   val queries: Queries
 
-  val maximumPoolSize: Option[Int]
+  val readerConnectionPool: DataSource
+
+  val writerConnectionPool: DataSource
 }
 
 object Database {
-  def apply(jdbcUrl: String): Database = {
+
+  // This *must* be 1 right now. We need to insert entries into the log in order; otherwise, we
+  // might end up dispatching (head + 2) before (head + 1), which will result in missing out an
+  // event when reading the log.
+  //
+  // To be able to process commits in parallel, we will need to fail reads and retry if there are
+  // entries missing.
+  private val MaximumWriterConnectionPoolSize: Int = 1
+
+  def apply(jdbcUrl: String)(implicit loggingContext: LoggingContext): Database = {
     jdbcUrl match {
-      case url if url.startsWith("jdbc:h2:") => new H2Database
-      case url if url.startsWith("jdbc:sqlite:") => new SqliteDatabase
+      case url if url.startsWith("jdbc:h2:") => new H2Database(jdbcUrl)
+      case url if url.startsWith("jdbc:sqlite:") => new SqliteDatabase(jdbcUrl)
       case _ => throw new InvalidDatabaseException(jdbcUrl)
     }
   }
 
-  final class H2Database extends Database {
+  final class H2Database(jdbcUrl: String)(implicit loggingContext: LoggingContext)
+      extends Database {
     override val queries: Queries = new H2Queries
 
-    override val maximumPoolSize: Option[Int] = None
+    override val readerConnectionPool: DataSource with Closeable =
+      newHikariDataSource(jdbcUrl, maximumPoolSize = None)
+
+    override val writerConnectionPool: DataSource with Closeable =
+      newHikariDataSource(jdbcUrl, maximumPoolSize = Some(MaximumWriterConnectionPoolSize))
+
+    ContextualizedLogger.get(this.getClass).info(s"Connected the ledger to $jdbcUrl.")
+
+    override def close(): Unit = {
+      readerConnectionPool.close()
+      writerConnectionPool.close()
+    }
   }
 
-  final class SqliteDatabase extends Database {
+  final class SqliteDatabase(jdbcUrl: String)(implicit loggingContext: LoggingContext)
+      extends Database {
+    private val connectionPool: DataSource with Closeable =
```

Comment:
I don't quite understand why `SqliteDatabase` shares the same connection pool with just 1 connection between reading and writing. We could still use a separate connection for reading the data for the `ReadService`, or is there a problem with that?

Reply:
Nope, the file gets "locked" when we're in the middle of a transaction and trying to access it concurrently results in an exception.

Reply:
Oh wow, that's ... unexpected. I thought it would "queue" up the write after the reads are done. Oh well.

Reply:
I think the JDBC driver is supposed to handle file system locks by retrying, but it doesn't right now.